File: nodejs-compressor/src/lexer/TokenType.js
/**
 * @fileoverview Token types for ASON 2.0 Lexer
 *
 * Defines all token types used in the ASON 2.0 format specification.
 * Each token type represents a distinct syntactic element in the language.
 *
 * @module TokenType
 * @license MIT
 * @version 2.0.0
 */

/**
 * Enumeration of all ASON 2.0 token types.
 *
 * @enum {string}
 * @readonly
 */
export const TokenType = Object.freeze({
  // Structural delimiters
  /** Section marker: @section_name */
  SECTION: '@',

  /** Key-value separator: key:value */
  COLON: ':',

  /** Field separator in tabular arrays: value1|value2 */
  PIPE: '|',

  /** Array item marker (YAML-style): - item */
  DASH: '-',

  /** Line continuation */
  BACKSLASH: '\\',

  // Brackets and braces
  /** Object start: { */
  LBRACE: '{',

  /** Object end: } */
  RBRACE: '}',

  /** Array start or count indicator: [ */
  LBRACKET: '[',

  /** Array end: ] */
  RBRACKET: ']',

  // References
  /** Named variable reference: $var_name */
  VAR_REF: '$',

  /** Object alias reference: &obj0 */
  OBJ_REF: '&',

  /** Numeric reference (legacy, deprecated): #0 */
  NUM_REF: '#',

  // Reserved keywords
  /** Definitions section: $def: */
  DEF_KEYWORD: '$def:',

  /** Data section: $data: */
  DATA_KEYWORD: '$data:',

  // Value types
  /** String value (quoted or unquoted) */
  STRING: 'STRING',

  /** Numeric value: 123, 45.67, -3.14 */
  NUMBER: 'NUMBER',

  /** Boolean value: true, false */
  BOOLEAN: 'BOOLEAN',

  /** Null value: null */
  NULL: 'NULL',

  /** Identifier (key name or reference name) */
  IDENTIFIER: 'IDENTIFIER',

  // Whitespace and formatting
  /** Newline character(s): \n, \r\n, \r */
  NEWLINE: 'NEWLINE',

  /** Indentation (spaces or tabs) */
  INDENT: 'INDENT',

  /** Whitespace within a line */
  WHITESPACE: 'WHITESPACE',

  // Comments
  /** Single-line comment: # comment */
  COMMENT: 'COMMENT',

  /** Multi-line comment start: #| */
  COMMENT_START: 'COMMENT_START',

  /** Multi-line comment end: |# */
  COMMENT_END: 'COMMENT_END',

  // Special markers
  /** Array count and schema marker: [N]{fields} */
  ARRAY_MARKER: 'ARRAY_MARKER',

  /** Schema definition marker: :schema{} */
  SCHEMA_MARKER: 'SCHEMA_MARKER',

  /** Dot for path notation: a.b.c */
  DOT: '.',

  /** Comma (for inline arrays/objects) */
  COMMA: ',',

  // Control
  /** End of file */
  EOF: 'EOF',

  /** Unknown/error token */
  ERROR: 'ERROR'
});

/**
 * Helper function to check if a token type represents a value.
 *
 * @param {string} type - Token type to check
 * @returns {boolean} True if token type represents a value
 *
 * @example
 * isValueType(TokenType.STRING) // true
 * isValueType(TokenType.NUMBER) // true
 * isValueType(TokenType.COLON) // false
 */
export function isValueType(type) {
  return [
    TokenType.STRING,
    TokenType.NUMBER,
    TokenType.BOOLEAN,
    TokenType.NULL,
    TokenType.IDENTIFIER
  ].includes(type);
}

/**
 * Helper function to check if a token type represents a reference.
 *
 * @param {string} type - Token type to check
 * @returns {boolean} True if token type represents a reference
 *
 * @example
 * isReferenceType(TokenType.VAR_REF) // true
 * isReferenceType(TokenType.OBJ_REF) // true
 * isReferenceType(TokenType.STRING) // false
 */
export function isReferenceType(type) {
  return [
    TokenType.VAR_REF,
    TokenType.OBJ_REF,
    TokenType.NUM_REF
  ].includes(type);
}

/**
 * Helper function to check if a token type represents a bracket.
 *
 * @param {string} type - Token type to check
 * @returns {boolean} True if token type is a bracket
 */
export function isBracketType(type) {
  return [
    TokenType.LBRACE,
    TokenType.RBRACE,
    TokenType.LBRACKET,
    TokenType.RBRACKET
  ].includes(type);
}

/**
 * Helper function to check if a token should be ignored during parsing.
 *
 * @param {string} type - Token type to check
 * @returns {boolean} True if token should be ignored
 */
export function isIgnorableType(type) {
  return [
    TokenType.COMMENT,
    TokenType.COMMENT_START,
    TokenType.COMMENT_END,
    TokenType.WHITESPACE
  ].includes(type);
}

/**
 * Get a human-readable name for a token type.
 *
 * @param {string} type - Token type
 * @returns {string} Human-readable name
 *
 * @example
 * getTokenTypeName(TokenType.VAR_REF) // "variable reference ($)"
 * getTokenTypeName(TokenType.COLON) // "colon (:)"
 */
export function getTokenTypeName(type) {
  const names = {
    [TokenType.SECTION]: 'section marker (@)',
    [TokenType.COLON]: 'colon (:)',
    [TokenType.PIPE]: 'pipe (|)',
    [TokenType.DASH]: 'dash (-)',
    [TokenType.VAR_REF]: 'variable reference ($)',
    [TokenType.OBJ_REF]: 'object reference (&)',
    [TokenType.NUM_REF]: 'numeric reference (#)',
    [TokenType.LBRACE]: 'left brace ({)',
    [TokenType.RBRACE]: 'right brace (})',
    [TokenType.LBRACKET]: 'left bracket ([)',
    [TokenType.RBRACKET]: 'right bracket (])',
    [TokenType.DOT]: 'dot (.)',
    [TokenType.COMMA]: 'comma (,)',
    [TokenType.STRING]: 'string',
    [TokenType.NUMBER]: 'number',
    [TokenType.BOOLEAN]: 'boolean',
    [TokenType.NULL]: 'null',
    [TokenType.IDENTIFIER]: 'identifier',
    [TokenType.NEWLINE]: 'newline',
    [TokenType.EOF]: 'end of file'
  };

  return names[type] || type;
}

File: python-compressor/src/ason/lexer/token_type.py
"""
Token types for ASON 2.0 Lexer

Defines all token types used in the ASON 2.0 format specification.
Each token type represents a distinct syntactic element in the language.

Module: token_type
License: MIT
Version: 2.0.0
"""

from enum import Enum
from typing import List


class TokenType(str, Enum):
    """Enumeration of all ASON 2.0 token types."""

    # Structural delimiters
    SECTION = '@'  # Section marker: @section_name
    COLON = ':'  # Key-value separator: key:value
    PIPE = '|'  # Field separator in tabular arrays: value1|value2
    DASH = '-'  # Array item marker (YAML-style): - item
    BACKSLASH = '\\'  # Line continuation

    # Brackets and braces
    LBRACE = '{'  # Object start: {
    RBRACE = '}'  # Object end: }
    LBRACKET = '['  # Array start or count indicator: [
    RBRACKET = ']'  # Array end: ]

    # References
    VAR_REF = '$'  # Named variable reference: $var_name
    OBJ_REF = '&'  # Object alias reference: &obj0
    NUM_REF = '#'  # Numeric reference (legacy, deprecated): #0

    # Reserved keywords
    DEF_KEYWORD = '$def:'  # Definitions section: $def:
    DATA_KEYWORD = '$data:'  # Data section: $data:

    # Value types
    STRING = 'STRING'  # String value (quoted or unquoted)
    NUMBER = 'NUMBER'  # Numeric value: 123, 45.67, -3.14
    BOOLEAN = 'BOOLEAN'  # Boolean value: true, false
    NULL = 'NULL'  # Null value: null
    IDENTIFIER = 'IDENTIFIER'  # Identifier (key name or reference name)

    # Whitespace and formatting
    NEWLINE = 'NEWLINE'  # Newline character(s): \n, \r\n, \r
    INDENT = 'INDENT'  # Indentation (spaces or tabs)
    WHITESPACE = 'WHITESPACE'  # Whitespace within a line

    # Comments
    COMMENT = 'COMMENT'  # Single-line comment: # comment
    COMMENT_START = 'COMMENT_START'  # Multi-line comment start: #|
    COMMENT_END = 'COMMENT_END'  # Multi-line comment end: |#

    # Special markers
    ARRAY_MARKER = 'ARRAY_MARKER'  # Array count and schema marker: [N]{fields}
    SCHEMA_MARKER = 'SCHEMA_MARKER'  # Schema definition marker: :schema{}
    DOT = '.'  # Dot for path notation: a.b.c
    COMMA = ','  # Comma (for inline arrays/objects)

    # Control
    EOF = 'EOF'  # End of file
    ERROR = 'ERROR'  # Unknown/error token


def is_value_type(token_type: TokenType) -> bool:
    """
    Check if a token type represents a value.

    Args:
        token_type: Token type to check

    Returns:
        True if token type represents a value

    Example:
        >>> is_value_type(TokenType.STRING)
        True
        >>> is_value_type(TokenType.COLON)
        False
    """
    return token_type in [
        TokenType.STRING,
        TokenType.NUMBER,
        TokenType.BOOLEAN,
        TokenType.NULL,
        TokenType.IDENTIFIER
    ]


def is_reference_type(token_type: TokenType) -> bool:
    """
    Check if a token type represents a reference.

    Args:
        token_type: Token type to check

    Returns:
        True if token type represents a reference

    Example:
        >>> is_reference_type(TokenType.VAR_REF)
        True
        >>> is_reference_type(TokenType.STRING)
        False
    """
    return token_type in [
        TokenType.VAR_REF,
        TokenType.OBJ_REF,
        TokenType.NUM_REF
    ]


def is_bracket_type(token_type: TokenType) -> bool:
    """
    Check if a token type represents a bracket.

    Args:
        token_type: Token type to check

    Returns:
        True if token type is a bracket
    """
    return token_type in [
        TokenType.LBRACE,
        TokenType.RBRACE,
        TokenType.LBRACKET,
        TokenType.RBRACKET
    ]


def is_ignorable_type(token_type: TokenType) -> bool:
    """
    Check if a token should be ignored during parsing.

    Args:
        token_type: Token type to check

    Returns:
        True if token should be ignored
    """
    return token_type in [
        TokenType.COMMENT,
        TokenType.COMMENT_START,
        TokenType.COMMENT_END,
        TokenType.WHITESPACE
    ]


def get_token_type_name(token_type: TokenType) -> str:
    """
    Get a human-readable name for a token type.

    Args:
        token_type: Token type

    Returns:
        Human-readable name

    Example:
        >>> get_token_type_name(TokenType.VAR_REF)
        'variable reference ($)'
    """
    names = {
        TokenType.SECTION: 'section marker (@)',
        TokenType.COLON: 'colon (:)',
        TokenType.PIPE: 'pipe (|)',
        TokenType.DASH: 'dash (-)',
        TokenType.VAR_REF: 'variable reference ($)',
        TokenType.OBJ_REF: 'object reference (&)',
        TokenType.NUM_REF: 'numeric reference (#)',
        TokenType.LBRACE: 'left brace ({)',
        TokenType.RBRACE: 'right brace (})',
        TokenType.LBRACKET: 'left bracket ([)',
        TokenType.RBRACKET: 'right bracket (])',
        TokenType.DOT: 'dot (.)',
        TokenType.COMMA: 'comma (,)',
        TokenType.STRING: 'string',
        TokenType.NUMBER: 'number',
        TokenType.BOOLEAN: 'boolean',
        TokenType.NULL: 'null',
        TokenType.IDENTIFIER: 'identifier',
        TokenType.NEWLINE: 'newline',
        TokenType.EOF: 'end of file'
    }

    return names.get(token_type, str(token_type))
